File: mobile-app/ios/Roost/Net/ApiClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking  // URLSession on Linux (the repo's Linux test harness)
#endif

/// Typed errors surfaced to the UI. `unauthorized` (401) drops to pairing;
/// `forbidden` (403) is a scope bug — show but stay paired (API.md §1).
enum ApiError: Error, Equatable {
    case unauthorized               // 401 — drop to pairing screen
    case forbidden(String)          // 403 — show, stay paired
    case notFound(String)           // 404
    case http(Int, String)          // other non-2xx with server detail
    case transport(String)          // URLSession / decode failure
}

/// Small async/await HTTP client against one control plane. Holds the base URL
/// and bearer token; nothing global. The stores own an instance and recreate it
/// on re-pair. Bytes streaming for SSE lives in `LogStream`, which borrows the
/// same `session` + auth.
struct ApiClient {
    let baseURL: URL
    let token: String
    let session: URLSession

    init(baseURL: URL, token: String, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.token = token
        self.session = session
    }

    // MARK: Request building

    func request(_ path: String, query: [URLQueryItem] = [],
                 method: String = "GET", body: Data? = nil) -> URLRequest {
        var comps = URLComponents(url: baseURL.appendingPathComponent(path),
                                  resolvingAgainstBaseURL: false)!
        if !query.isEmpty { comps.queryItems = query }
        var req = URLRequest(url: comps.url!)
        req.httpMethod = method
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        if let body {
            req.httpBody = body
            req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        return req
    }

    /// Map an HTTP response to a typed error, decoding the `{"detail": …}`
    /// envelope when present (API.md §1).
    static func mapError(status: Int, data: Data) -> ApiError {
        let detail = (try? JSONDecoder().decode(ErrorEnvelope.self, from: data))?.detail
            ?? String(data: data, encoding: .utf8) ?? "HTTP \(status)"
        switch status {
        case 401: return .unauthorized
        case 403: return .forbidden(detail)
        case 404: return .notFound(detail)
        default: return .http(status, detail)
        }
    }

    private func sendData(_ req: URLRequest) async throws -> Data {
        let data: Data
        let resp: URLResponse
        do {
            (data, resp) = try await session.data(for: req)
        } catch {
            throw ApiError.transport(error.localizedDescription)
        }
        guard let http = resp as? HTTPURLResponse else {
            throw ApiError.transport("non-HTTP response")
        }
        guard (200...299).contains(http.statusCode) else {
            throw Self.mapError(status: http.statusCode, data: data)
        }
        return data
    }

    private func send<T: Decodable>(_ req: URLRequest, as type: T.Type) async throws -> T {
        let data = try await sendData(req)
        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw ApiError.transport("decode: \(error)")
        }
    }

    // MARK: Endpoints (only those in API.md)

    /// Unauthenticated reachability probe. `timeout` is the per-request deadline;
    /// the pairing flow passes a short one (`Self.pairingProbeSession`) so a dead
    /// host fails fast instead of the old ~30 s silent wait (R83).
    static func healthz(baseURL: URL, session: URLSession = .shared,
                        timeout: TimeInterval = 8) async throws -> Healthz {
        var req = URLRequest(url: baseURL.appendingPathComponent("healthz"))
        req.timeoutInterval = timeout
        let data: Data, resp: URLResponse
        do { (data, resp) = try await session.data(for: req) }
        catch { throw ApiError.transport(error.localizedDescription) }
        guard let http = resp as? HTTPURLResponse, (200...299).contains(http.statusCode)
        else { throw ApiError.transport("healthz unreachable") }
        do { return try JSONDecoder().decode(Healthz.self, from: data) }
        catch { throw ApiError.transport("healthz decode") }
    }

    /// Short, fail-fast timeout for the *pairing* reachability probe (R83). The
    /// normal app/SSE session (`AppState.makeClient`) keeps its long-lived
    /// timeouts — log streams stay open for minutes — so the short deadline is
    /// scoped to this one throwaway probe and never touches steady-state traffic.
    ///
    /// 5 s: a control plane on the same LAN answers `/healthz` in tens of ms; a
    /// dead host on-network (the R83 case) is detectable in well under 5 s, and a
    /// human typing on a phone reads a ~5 s "couldn't reach it" as snappy, not
    /// broken — whereas the old ~30 s read as a hang. `waitsForConnectivity`
    /// stays false (the default) so an unreachable host fails at the deadline
    /// rather than parking the request until the network changes.
    static let pairingProbeTimeout: TimeInterval = 5

    /// A dedicated session for the pairing probe with the short deadline baked in
    /// at the config level (belt-and-suspenders with the per-request timeout) and
    /// connectivity-waiting explicitly off, so a dead host can never hang it.
    static func makePairingProbeSession() -> URLSession {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = pairingProbeTimeout
        config.timeoutIntervalForResource = pairingProbeTimeout
        #if !canImport(FoundationNetworking)
        // Apple platforms only: keep connectivity-waiting off (its default) so an
        // unreachable host fails at the deadline rather than parking the request
        // until the network changes. On Linux's FoundationNetworking this property
        // is get-only (and already false), so we leave it alone there.
        config.waitsForConnectivity = false
        #endif
        return URLSession(configuration: config)
    }

    func derived(limit: Int = 40) async throws -> Derived {
        try await send(request("derived", query: [.init(name: "limit", value: String(limit))]),
                       as: Derived.self)
    }

    /// Raw body so the caller can also feed the offline cache (one fetch, one decode).
    func derivedRaw(limit: Int = 40) async throws -> Data {
        try await sendData(request("derived", query: [.init(name: "limit", value: String(limit))]))
    }

    /// Fleet list (API.md §2a): the full worker rows — the same rows
    /// `/derived` embeds, served under the §1 client permission set.
    func workers() async throws -> [Worker] {
        try await send(request("workers"), as: [Worker].self)
    }

    func job(_ id: String) async throws -> Job {
        try await send(request("jobs/\(id)"), as: Job.self)
    }

    func jobDerived(_ id: String) async throws -> Run {
        try await send(request("jobs/\(id)/derived"), as: Run.self)
    }

    func logs(_ id: String, since: Int, limit: Int = 1000) async throws -> LogPage {
        try await send(request("jobs/\(id)/logs", query: [
            .init(name: "since", value: String(since)),
            .init(name: "limit", value: String(limit)),
        ]), as: LogPage.self)
    }

    func tree(_ id: String) async throws -> [Job] {
        try await send(request("jobs/\(id)/tree"), as: [Job].self)
    }

    func submit(_ submit: JobSubmit) async throws -> Job {
        let body = try JSONEncoder().encode(submit)
        return try await send(request("jobs", method: "POST", body: body), as: Job.self)
    }

    @discardableResult
    func cancel(_ id: String, tree: Bool = false) async throws -> CancelResponse {
        let query = tree ? [URLQueryItem(name: "tree", value: "true")] : []
        return try await send(request("jobs/\(id)", query: query, method: "DELETE"),
                              as: CancelResponse.self)
    }

    // MARK: Follow-up input (R38, API.md §4)

    /// Queue a follow-up message for a RUNNING job. The CP returns
    /// `{input_id, job_id, state:"queued"}`; the owning worker delivers it on its
    /// heartbeat. A terminal job is rejected `.http(409, …)`, empty text 400,
    /// >64 KiB 413 — all surfaced as typed `ApiError` with the server detail.
    func sendInput(_ id: String, text: String) async throws -> JobInputAck {
        let body = try JSONEncoder().encode(JobInputSubmit(text: text))
        return try await send(request("jobs/\(id)/input", method: "POST", body: body),
                              as: JobInputAck.self)
    }

    /// Poll a job's follow-up queue for the delivery outcome (API.md §4).
    func inputs(_ id: String) async throws -> JobInputs {
        try await send(request("jobs/\(id)/inputs"), as: JobInputs.self)
    }

    // MARK: Publish (API.md §6)

    /// Stage a site bundle (raw tar.gz body) — publish step 1.
    func uploadBlob(name: String, data: Data) async throws -> BlobUploadResponse {
        var req = request("blobs", query: [.init(name: "name", value: name)],
                          method: "POST")
        req.httpBody = data
        req.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        return try await send(req, as: BlobUploadResponse.self)
    }

    /// Publish a staged bundle — step 2. `name` optional (defaults server-side
    /// to the blob name minus its tar suffix, then slugified).
    func publish(blobId: String, name: String? = nil) async throws -> Site {
        var payload = ["blob_id": blobId]
        if let name { payload["name"] = name }
        let body = try JSONEncoder().encode(payload)
        return try await send(request("publish", method: "POST", body: body),
                              as: Site.self)
    }

    /// One-shot publish (API.md §6a): the `tar.gz` IS the body, no staged blob.
    /// `name` is REQUIRED (no blob name to default from) and slugified server-side.
    /// Returns the same `Site` as the two-step `publish(blobId:)`.
    func publishBundle(name: String, data: Data) async throws -> Site {
        var req = request("publish", query: [.init(name: "name", value: name)],
                          method: "POST")
        req.httpBody = data
        // Non-JSON Content-Type selects the one-shot path server-side.
        req.setValue("application/gzip", forHTTPHeaderField: "Content-Type")
        return try await send(req, as: Site.self)
    }

    func sites() async throws -> [Site] {
        try await send(request("publish"), as: [Site].self)
    }

    // MARK: Schedules (API.md §7)

    /// List interval schedules, newest first.
    func schedules() async throws -> [Schedule] {
        try await send(request("schedules"), as: [Schedule].self)
    }

    /// Create an interval schedule. `every` is seconds or "<N>[smhd]" (e.g. "30m";
    /// 30 s floor server-side). `spec` follows the §3 submit shape (root jobs only).
    /// Returns the new `Schedule`; its first run fires one interval from now.
    func createSchedule(spec: [String: JSONValue], every: String,
                        name: String? = nil, enabled: Bool = true) async throws -> Schedule {
        let body = try JSONEncoder().encode(
            ScheduleCreate(spec: spec, every: every, name: name, enabled: enabled))
        return try await send(request("schedules", method: "POST", body: body),
                              as: Schedule.self)
    }

    /// Enable/disable a schedule (API.md §7c). Re-enabling restarts the clock
    /// server-side. Returns the updated `Schedule`.
    @discardableResult
    func setScheduleEnabled(_ id: String, enabled: Bool) async throws -> Schedule {
        let body = try JSONEncoder().encode(SchedulePatch(enabled: enabled))
        return try await send(request("schedules/\(id)", method: "PATCH", body: body),
                              as: Schedule.self)
    }

    /// Delete a schedule (API.md §7d) → `{"deleted": true, "id": …}`.
    @discardableResult
    func deleteSchedule(_ id: String) async throws -> ScheduleDeleteResponse {
        try await send(request("schedules/\(id)", method: "DELETE"),
                       as: ScheduleDeleteResponse.self)
    }
}

File: mobile-app/android/app/src/main/java/oss/roost/mobile/net/ApiClient.kt
package oss.roost.mobile.net

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import oss.roost.mobile.model.CancelAck
import oss.roost.mobile.model.Derived
import oss.roost.mobile.model.Healthz
import oss.roost.mobile.model.Job
import oss.roost.mobile.model.JobInputAck
import oss.roost.mobile.model.JobInputs
import oss.roost.mobile.model.LogPage
import oss.roost.mobile.model.Parsers
import oss.roost.mobile.model.Run
import oss.roost.mobile.model.Schedule
import oss.roost.mobile.model.Site
import oss.roost.mobile.model.StagedBlob
import oss.roost.mobile.model.Worker
import org.json.JSONObject
import java.io.BufferedReader
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

/**
 * The single networking surface. Every method is a suspend function that runs the blocking
 * HttpURLConnection call on Dispatchers.IO (API.md transport: plain HTTP to the CP URL).
 *
 * WHY HttpURLConnection and not OkHttp/Retrofit: the weight budget (DESIGN §7) — zero
 * third-party runtime deps. HUC + org.json + coroutines cover the whole contract.
 *
 * Auth/error model (API.md §1): every request carries Bearer <token>. A 401 throws
 * ApiException(401) which the app turns into "unpair"; 403 throws ApiException(403) which
 * stays paired and shows the detail. The SSE stream is handled separately (SseClient).
 */
class ApiClient(
    @Volatile var baseUrl: String,
    @Volatile var token: String?,
) {
    /** Thrown for any non-2xx with the parsed `{detail}` envelope. */
    class ApiException(val status: Int, val detail: String) : IOException("HTTP $status: $detail")

    // ---- public endpoints --------------------------------------------------------

    /** Unauthenticated reachability probe — used before accepting a pairing. */
    suspend fun healthz(): Healthz =
        Parsers.parseHealthz(getText("/healthz", auth = false))

    /** Raw body so the caller can also feed the offline cache (one fetch, one parse). */
    suspend fun derivedRaw(limit: Int = 40): String = getText("/derived?limit=$limit")

    suspend fun derived(limit: Int = 40): Derived =
        Parsers.parseDerived(derivedRaw(limit))

    suspend fun workers(): List<Worker> =
        Parsers.parseWorkers(getText("/workers"))

    suspend fun job(id: String): Job =
        Parsers.parseJob(getText("/jobs/${enc(id)}"))

    suspend fun jobStory(id: String): Run =
        Parsers.parseRunStory(getText("/jobs/${enc(id)}/derived"))

    suspend fun logs(id: String, since: Int, limit: Int = 1000): LogPage =
        Parsers.parseLogPage(getText("/jobs/${enc(id)}/logs?since=$since&limit=$limit"))

    suspend fun tree(id: String): List<Job> =
        Parsers.parseTree(getText("/jobs/${enc(id)}/tree"))

    suspend fun cancel(id: String, cascade: Boolean = false): CancelAck =
        Parsers.parseCancel(
            requestText("DELETE", "/jobs/${enc(id)}${if (cascade) "?tree=true" else ""}", null)
        )

    /**
     * Queue a follow-up message for a RUNNING job (R38, API.md §4). The CP returns
     * `{input_id, job_id, state:"queued"}`; the owning worker delivers it on its
     * heartbeat. A terminal job is rejected 409, empty text 400, >64 KiB 413 — all
     * surfaced as [ApiException] with the server detail.
     */
    suspend fun sendInput(id: String, text: String): JobInputAck =
        Parsers.parseJobInputAck(
            requestText("POST", "/jobs/${enc(id)}/input",
                JSONObject().put("text", text).toString()))

    /** Poll a job's follow-up queue for delivery outcome (API.md §4). */
    suspend fun inputs(id: String): JobInputs =
        Parsers.parseJobInputs(getText("/jobs/${enc(id)}/inputs"))

    /**
     * Submit a job (API.md §3). Sends ONLY the contract fields. `requires` is {} for
     * auto-place or {"worker": id} to pin. Returns the full job object; caller navigates
     * to job.id immediately.
     */
    suspend fun submit(
        intent: String,
        kind: String,            // "claude" | "command"
        pinWorker: String?,
        command: String?,
    ): Job {
        val body = JSONObject()
        body.put("kind", kind)
        val requires = JSONObject()
        if (pinWorker != null) requires.put("worker", pinWorker)
        body.put("requires", requires)
        if (kind == "command") {
            body.put("command", command ?: intent)
        } else {
            body.put("intent", intent)
            // can_dispatch makes the worker inject the roost MCP, so the agent
            // can SEE the fleet ("how many machines?") and spawn sub-jobs under
            // the existing depth/tree-budget guardrails. Without it the agent
            // is fleet-blind (API.md §3).
            body.put("hierarchy", JSONObject().put("can_dispatch", true))
        }
        return Parsers.parseJob(requestText("POST", "/jobs", body.toString()))
    }

    // ---- publish (API.md §6) -------------------------------------------------------

    /** Stage a site bundle (raw tar.gz bytes) — publish step 1. */
    suspend fun uploadBlob(name: String, bytes: ByteArray): StagedBlob =
        withContext(Dispatchers.IO) {
            val conn = open("/blobs?name=${enc(name)}")
            try {
                conn.requestMethod = "POST"
                conn.connectTimeout = 10_000
                conn.readTimeout = 60_000   // bundles are bigger than JSON bodies
                token?.let { conn.setRequestProperty("Authorization", "Bearer $it") }
                conn.setRequestProperty("Content-Type", "application/octet-stream")
                conn.doOutput = true
                conn.outputStream.use { it.write(bytes) }
                val code = conn.responseCode
                if (code in 200..299) {
                    Parsers.parseBlob(readBody(conn) ?: "")
                } else {
                    val err = readError(conn) ?: ""
                    throw Parsers.parseError(code, err)
                        .let { ApiException(it.status, it.detail) }
                }
            } finally {
                conn.disconnect()
            }
        }

    /**
     * Publish a staged bundle — step 2. `name` optional (server defaults it to the
     * blob name minus its tar suffix, then slugifies).
     */
    suspend fun publish(blobId: String, name: String? = null): Site {
        val body = JSONObject().put("blob_id", blobId)
        if (name != null) body.put("name", name)
        return Parsers.parseSite(requestText("POST", "/publish", body.toString()))
    }

    /**
     * One-shot publish (API.md §6a): the tar.gz IS the body, no staged blob.
     * `name` is required (no blob name to default from) and slugified server-side.
     * Returns the same [Site] as the two-step [publish]. A non-JSON Content-Type
     * selects the one-shot path server-side.
     */
    suspend fun publishBundle(name: String, bytes: ByteArray): Site =
        withContext(Dispatchers.IO) {
            val conn = open("/publish?name=${enc(name)}")
            try {
                conn.requestMethod = "POST"
                conn.connectTimeout = 10_000
                conn.readTimeout = 60_000   // bundles are bigger than JSON bodies
                token?.let { conn.setRequestProperty("Authorization", "Bearer $it") }
                conn.setRequestProperty("Content-Type", "application/gzip")
                conn.doOutput = true
                conn.outputStream.use { it.write(bytes) }
                val code = conn.responseCode
                if (code in 200..299) {
                    Parsers.parseSite(readBody(conn) ?: "")
                } else {
                    val err = readError(conn) ?: ""
                    throw Parsers.parseError(code, err)
                        .let { ApiException(it.status, it.detail) }
                }
            } finally {
                conn.disconnect()
            }
        }

    suspend fun sites(): List<Site> = Parsers.parseSites(getText("/publish"))

    // ---- schedules (API.md §7) -----------------------------------------------------

    /** List interval schedules, newest first. */
    suspend fun schedules(): List<Schedule> =
        Parsers.parseSchedules(getText("/schedules"))

    /**
     * Create an interval schedule (API.md §7a). `every` is seconds or "<N>[smhd]"
     * (e.g. "30m"; 30 s floor server-side). `spec` follows the §3 submit shape
     * (root jobs only). Returns the new [Schedule]; first run is one interval out.
     */
    suspend fun createSchedule(
        spec: Map<String, Any?>,
        every: String,
        name: String? = null,
        enabled: Boolean = true,
    ): Schedule {
        val body = JSONObject()
        body.put("spec", JSONObject(spec))
        body.put("every", every)
        body.put("enabled", enabled)
        if (name != null) body.put("name", name)
        return Parsers.parseSchedule(requestText("POST", "/schedules", body.toString()))
    }

    /**
     * Enable/disable a schedule (API.md §7c). Re-enabling restarts the clock
     * server-side. Returns the updated [Schedule].
     */
    suspend fun setScheduleEnabled(id: String, enabled: Boolean): Schedule {
        val body = JSONObject().put("enabled", enabled)
        return Parsers.parseSchedule(
            requestText("PATCH", "/schedules/${enc(id)}", body.toString()))
    }

    /** Delete a schedule (API.md §7d). 404 if unknown → ApiException(404). */
    suspend fun deleteSchedule(id: String): Boolean {
        val json = requestText("DELETE", "/schedules/${enc(id)}", null)
        return JSONObject(json).optBoolean("deleted", false)
    }

    // ---- HTTP plumbing -----------------------------------------------------------

    private suspend fun getText(path: String, auth: Boolean = true): String =
        requestText("GET", path, null, auth)

    /**
     * Run one request and return the body. Throws ApiException on non-2xx with the parsed
     * detail. All blocking I/O is confined to Dispatchers.IO.
     */
    private suspend fun requestText(
        method: String,
        path: String,
        body: String?,
        auth: Boolean = true,
    ): String = withContext(Dispatchers.IO) {
        val conn = open(path)
        try {
            conn.requestMethod = method
            conn.connectTimeout = 10_000
            conn.readTimeout = 20_000
            if (auth) token?.let { conn.setRequestProperty("Authorization", "Bearer $it") }
            conn.setRequestProperty("Accept", "application/json")
            if (body != null) {
                conn.doOutput = true
                conn.setRequestProperty("Content-Type", "application/json")
                conn.outputStream.use { it.write(body.toByteArray(Charsets.UTF_8)) }
            }
            val code = conn.responseCode
            if (code in 200..299) {
                readBody(conn) ?: ""
            } else {
                val err = readError(conn) ?: ""
                throw Parsers.parseError(code, err).let { ApiException(it.status, it.detail) }
            }
        } finally {
            conn.disconnect()
        }
    }

    /** Expose a configured connection for the SSE client without duplicating header logic. */
    fun openAuthed(path: String): HttpURLConnection {
        val conn = open(path)
        token?.let { conn.setRequestProperty("Authorization", "Bearer $it") }
        return conn
    }

    private fun open(path: String): HttpURLConnection {
        val full = if (path.startsWith("http")) path else baseUrl.trimEnd('/') + path
        return (URL(full).openConnection() as HttpURLConnection)
    }

    private fun readBody(conn: HttpURLConnection): String? =
        conn.inputStream?.bufferedReader()?.use(BufferedReader::readText)

    private fun readError(conn: HttpURLConnection): String? =
        conn.errorStream?.bufferedReader()?.use(BufferedReader::readText)

    private fun enc(s: String): String = URLEncoder.encode(s, "UTF-8")
}
